File: moonpoison/linked_list/2095.cpp
#include<bits/stdc++.h>
using namespace std;
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) 
    {
        ListNode ans(0, head);
        ListNode *p1=&ans;
        ListNode *p2=&ans;
        while(p1->next!=nullptr&&p1->next->next!=nullptr)
        {
            p2=p2->next;
            p1=p1->next->next;
        }
        p2->next=p2->next->next;
        return ans.next;
    }
};

File: ice0208/linked-list/2095.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head):
        # Linked List의 길이 구하기
        len = 1
        cursor = head
        while cursor.next:
            len += 1
            cursor = cursor.next

        # 길이가 1일 때, None을 return
        if len == 1:
            return None

        # middle의 index 구하기
        middle_index = len // 2

        # cursor prev -> cursor next 로 연결해서 middle 값을 제거
        cursor_index = 0
        cursor = head
        cursor_prev = None
        while cursor_index < middle_index:
            cursor_index += 1
            cursor_prev = cursor
            cursor = cursor.next

        middle_prev_cursor = cursor_prev
        middle_cursor = cursor

        # middle 노드 연결 끊기
        middle_prev_cursor.next = middle_cursor.next
        middle_cursor.next = None

        return head
